### ISSUE TITLE
implement importance scores for random forest

### ISSUE THREAD

**Issue by lacava** [lacava/feat#220]
first referenced in issue #219

**Comment by lacava** [comment]
basic gameplan:

 - copy RandomCARTree and RandomForest from shogun into src/ml/shogun/ as My*
 - copy the feature_importances() method from MyCARTree into RandomCARTree 
 - make a feature_importances() method in RandomForest that takes the average of the RandomCARTree feature_importances() vector across all trees, resulting in w of size num_features
- return feature_importances() from RandomForest in a similar way to CART in the ML::get_weights() class

**Comment by lacava** [comment]
ok, some detail updates:

- CMyRandomForest
  - inherit from CRandomForest
  - any function where CRandomCARTree is called needs to be rewritten with calls to CMyRandomCARTree
  - implements feature_importances()

- CMyRandomCARTree
  - inherits from CMyCARTree
  - starts with copy and paste of CRandomCARTree and changes compute_best_attributes()
  - wherever CCARTree is called, call CMyCARTree
  - (should inherit feature_importance() from CMyCARTree)

**Comment by lacava** [comment]
OK! getting close. now we need a set_probabilities function that looks something like:

```
 CBinaryLabels* CMyRandomForest::set_probabilities(CLabels* labels, CFeatures* data)
 {
     SGMatrix<float64_t> output = apply_outputs_without_combination(data);
 
     CMeanRule* mean_rule = new CMeanRule();

     SGVector<float64_t> probabilities = mean_rule->combine(output);
    
     labels->set_values(probabilities);

     SG_UNREF(mean_rule);
 
 }
 ```

that function should be called in ML::fit (where it is called for CART) and ML::predict

**Comment by tilakhere** [comment]
Major Updates for fixing this

- importing all files for custom RandomForest Implementation - commit https://github.com/lacava/feat/commit/c6330952e0add979bcfe0bc8eb4ce83510faffc4
- change how custom RandomCARTree Objects are cloned - commit https://github.com/lacava/feat/commit/faa89bce713a8611ec93aa5f72d78474c7f0fea1
- handling issue when all features are zero and shogun throws a segfault - commit https://github.com/lacava/feat/commit/671c856343104d79d58b4c913d23c04c7e695616
- implementing feature importance scores and setting probabilities - commit https://github.com/lacava/feat/commit/8009666f0099dbdc9df18fcb9358060048d272ed and 6699d199fb52b6750d005896d937c0bc0e31825e

**Comment by lacava** [comment]
closed as of PR #224
